command/certificate/handshake.go

```diff
@@ -0,0 +1,260 @@
+			},
+			cli.BoolFlag{
+				Name: "insecure",
+				Usage: `Use an insecure client to retrieve a remote peer certificate. Useful for
+debugging invalid certificates remotely.`,
+			},
+		},
+	}
+}
+
+func handshakeAction(ctx *cli.Context) error {
+	if err := errs.NumberOfArguments(ctx, 1); err != nil {
+		return err
+	}
+
+	var (
+		addr         = ctx.Args().First()
+		tlsVersion   = ctx.String("tls")
+		roots        = ctx.String("roots")
+		serverName   = ctx.String("servername")
+		certFile     = ctx.String("cert")
+		keyFile      = ctx.String("key")
+		passwordFile = ctx.String("password-file")
+		printChains  = ctx.Bool("chain")
+		printPeer    = ctx.Bool("peer")
+		insecure     = ctx.Bool("insecure")
+		rootCAs      *x509.CertPool
+		err          error
+	)
+
+	switch {
+	case certFile != "" && keyFile == "":
+		return errs.RequiredWithFlag(ctx, "cert", "key")
+	case keyFile != "" && certFile == "":
+		return errs.RequiredWithFlag(ctx, "key", "cert")
+	}
+
+	// Parse address
+	if u, ok, err := trimURL(addr); err != nil {
+		return err
+	} else if ok {
+		addr = u
+	}
+	if _, _, err := net.SplitHostPort(addr); err != nil {
+		addr = net.JoinHostPort(addr, "443")
+	}
+
+	// Load certificate and if
+	var certificates []tls.Certificate
+	if certFile != "" && keyFile != "" {
+		opts := []pemutil.Options{}
+		if passwordFile != "" {
+			opts = append(opts, pemutil.WithPasswordFile(passwordFile))
+		}
+		crt, err := cryptoutil.LoadTLSCertificate(certFile, keyFile, opts...)
+		if err != nil {
+			return err
+		}
+		certificates = []tls.Certificate{crt}
+	}
+
+	// Get the list of roots used to validate the certificate.
+	if roots != "" {
+		rootCAs, err = x509util.ReadCertPool(roots)
+		if err != nil {
+			return fmt.Errorf("error loading root certificate pool from %q: %w", roots, err)
+		}
+	} else {
+		rootCAs, err = x509.SystemCertPool()
+		if err != nil {
+			return fmt.Errorf("error loading the system cert pool: %w", err)
+		}
+	}
+
+	// Get the tls version to use. Defaults to TLS 1.2+
+	minVersion, maxVersion, err := getTLSVersions(tlsVersion)
+	if err != nil {
+		return err
+	}
+
+	tlsConfig := &tls.Config{
+		MinVersion:         minVersion,
+		MaxVersion:         maxVersion,
+		RootCAs:            rootCAs,
+		InsecureSkipVerify: insecure,
+		ServerName:         serverName,
+		Certificates:       certificates,
+	}
+
+	cs, err := tlsDialWithFallback(addr, tlsConfig)
+	if err != nil {
+		return err
+	}
+
+	// Print only the list of verified chains
+	if printChains {
+		for _, chain := range cs.VerifiedChains {
+			for _, crt := range chain {
+				fmt.Print(string(pem.EncodeToMemory(&pem.Block{
+					Type:  "CERTIFICATE",
+					Bytes: crt.Raw,
+				})))
+			}
+		}
+		return nil
+	}
+
+	// Print only the peer certificates
+	if printPeer {
+		for _, crt := range cs.PeerCertificates {
+			fmt.Print(string(pem.EncodeToMemory(&pem.Block{
+				Type: "CERTIFICATE", Bytes: crt.Raw,
+			})))
+		}
+		return nil
+	}
```

Comment:
Maybe needs an option to continue down, so that the connection details are shown in addition to these too? Or make that the default, instead of returning early?

Reply:
I'm not sure about this. I make them independent commands because a certificate, especially if you include all the intermediates, is quite noisy. I also wanted to pipe those certificates to another command.

Reply:
That's fair, but in practice providing one of the `printPeer` or `printChains` options is doing the same as `step certificate inspect https://example.com` and `step certificate inspect https://example.com --bundle`, respectively. 

If instead execution would continue after printing the details, it would be a bit more like `curl -v` (with more certificate details, of course).